bpo-41543: contextlib.nullcontext can fill in for an async context manager

tomgrin10 opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->

Make `contextlib.nullcontext` implement `AbstractAsyncContextManager`, so it can be used as a fill in for async context mangers.

<!-- issue-number: [bpo-41543](https://bugs.python.org/issue41543) -->
https://bugs.python.org/issue41543
<!-- /issue-number -->

iritkatriel reviewed (commented):
Missing a test

iritkatriel commented on Doc/library/contextlib.rst:
```diff
@@ -185,6 +185,19 @@ Functions and classes provided:
+       async def send_http(session=None):
+          if session:
+              # If no http session, create it with aiohttp
+              cm = aiohttp.ClientSession()
```
Should this be "if not session"  instead of "if session"?

iritkatriel reviewed (commented):
Still need to squash the commits.

1st1 commented on Doc/library/contextlib.rst:
```diff
@@ -185,6 +185,19 @@ Functions and classes provided:
           with cm as file:
               # Perform processing on the file
 
+   It can also be used as a stand-in for :ref:`asynchronous context managers <async-context-managers>`::
```
The line max length is 79 characters, please reformat

tomgrin10 replied:
Fixed

tomgrin10 committed:
bpo-41543: contextlib.nullcontext can fill in for an async context manager

Add NEWS.d blurb


Add tests


Fix test


Update nullcontext docs


Fix example in docs


Format lines

asvetlov reviewed (approved):
LGTM.

As the aiohttp author, I never use such approach :)

asvetlov reviewed (changes requested):
Need `..versionchanged:: 3.10` directive in docs that says 'added support for async context manager' or something like this

asvetlov commented on Doc/library/contextlib.rst:
```diff
@@ -218,6 +218,20 @@ Functions and classes provided:
+          async with cm as session:
+              # Send http requests with session
+
    .. versionadded:: 3.7
```
```suggestion
   .. versionadded:: 3.7

   .. versionchanged:: 3.10
      :term:`asynchronous context manager` support was added.

```

asvetlov committed:
Update Doc/library/contextlib.rst
